File: scala/src/main/scala/com/shekhargulati/ninetynine_problems/_02_arithmetic/P36.scala
package com.shekhargulati.ninetynine_problems._02_arithmetic

/**
  * (**) A list of Goldbach compositions.
  */
object P36 {

  def goldbach_list(range: Range): List[(Int, (Int, Int))] = {
    range.filter(_ % 2 == 0).filter(_ > 2).map(even => {
      val g = P35.goldbach(even)
      (even, (g.head, g.last))
    }).toList
  }

  def goldbach_list1(range: Range, gt: Int): List[(Int, (Int, Int))] = {
    goldbach_list(range).filter(g => g._2._1 > gt && g._2._2 > gt)
  }


}

File: java8/src/main/java/com/shekhargulati/ninetynine_problems/_02_arithmetic/P36.java
package com.shekhargulati.ninetynine_problems._02_arithmetic;

import java.util.AbstractMap.SimpleEntry;
import java.util.List;
import java.util.stream.IntStream;

import static java.util.stream.Collectors.toList;

/**
 * (**) A list of Goldbach compositions.
 */
public class P36 {

    public static List<SimpleEntry<Integer, List<Integer>>> goldbach_list(IntStream range) {
        return range
                .filter(n -> n % 2 == 0)
                .filter(even -> even > 2)
                .mapToObj(even -> new SimpleEntry<>(even, P35.goldbach(even)))
                .collect(toList());
    }

    public static List<SimpleEntry<Integer, List<Integer>>> goldbach_list1(IntStream range, int greaterThan) {
        return goldbach_list(range)
                .stream()
                .filter(g -> g.getValue().get(0) > greaterThan && g.getValue().get(1) > greaterThan)
                .collect(toList());

    }
}
